File: Java/WoM_T3/src/main/java/sep3/service/interfaces/ICowService.java
package sep3.service.interfaces;
import sep3.dto.MilkDtos;
import sep3.dto.cowDTO.CowCreationDTO;
import sep3.dto.cowDTO.CowDataDTO;
import java.util.List;

public interface ICowService {
  CowDataDTO addCow(CowCreationDTO cow);
  CowDataDTO getCowById(long cowToFindId);
  CowDataDTO getCowByRegNo(String regNo);
  List<CowDataDTO> getCowsByDepartmentId(long departmentId);
  List<MilkDtos.MilkDto> getCowMilk(long cowId);
  List<CowDataDTO> getAllCows();
  CowDataDTO updateCow(CowDataDTO changesToCow, long userId);
  void updateManyCowsHealth(List<Long> cowsIds, boolean healthUpdate, long userId);
  void deleteCow(long id);
}

File: C_Sharp/Server/WoM_WebApi/Services/Interfaces/ICowService.cs

using ApiContracts;

namespace WoM_WebApi.Services.Interfaces;

// Service: cow CRUD + queries
public interface ICowService
{
    Task<CowDto> CreateAsync(CowCreationDto dto);
    Task<CowDto> GetByIdAsync(long id);
    Task<IEnumerable<CowDto>> GetAllAsync();
    Task<CowDto> GetByRegNoAsync(string regNo);

    // Updates
    Task<CowDto> UpdateCowAsync(CowDto dto, long requesterUserId);
    IAsyncEnumerable<CowDto> UpdateBatchAsync(IEnumerable<CowDto> dtos, long requesterUserId);
    IAsyncEnumerable<CowDto> UpdateCowsHealthAsync(IEnumerable<long> cowIds, bool healthUpdate, long requesterUserId);

    // Deletes
    Task DeleteAsync(long id);
    Task DeleteBatchAsync(long[] ids);

    //other methods
    Task<IEnumerable<CowDto>> GetCowsByDepartmentAsync(long departmentId);
    Task<IEnumerable<MilkDto>> GetMilkByCowIdAsync(long cowId);
}
